### ISSUE TITLE
Atrous Convolution

### ISSUE THREAD

**Issue by chenxiaodanhit** [tensorpack/tensorpack#1110]
Hi  ppwwyyxx,
    Sorry to bother you, I countered two problems recently when i want to  improve rpn.

    1.when i modify the convolution of rpn into atrous convolution with padding is equals to same and dilation_rate is 2, the rpn_losses went wrong.This is where things go wrong:                                               **valid_label_logits = tf.boolean_mask(label_logits, valid_mask)**
    The error message is as follows:
**InvalidArgumentError (see above for traceback): Input to reshape is a tensor with 11400 values, but the requested shape has 196608**

  2.I tried my best but I cannot print the value of tensor in model_rpn.py
  
  Hope to get your advice,Thank you!

**Comment by ppwwyyxx** [comment]
1. Without runnable code there is nothing I can tell from this error message. It's likely that your modified code produces incompatible shape with other tensors.

For anyone to better understand and diagnose your issue, please post relevant details following the issue template (click "New Issue" -> "Unexpected Problems / Bugs", or visit [this link to post issues about unexpected problems](https://github.com/tensorpack/tensorpack/issues/new?template=unexpected-problems---bugs.md))

2. I'm not sure I understand what that means. If you just want to print tensor values during training, it is in [the FAQ](https://tensorpack.readthedocs.io/tutorial/faq.html#how-to-print-dump-intermediate-results-in-training)

**Comment by chenxiaodanhit** [comment]
Hi ppwwyyxx,
  1.In the training pharse, I only modified dilation_rate in model_rpn.py(rpn_head function): hidden = Conv2D('conv0', featuremap, channel, 3, dilation_rate=2, activation=tf.nn.relu, padding='SAME').But it went wrong in  ### valid_label_logits = tf.boolean_mask(label_logits, valid_mask)### in model_rpn.py(rpn_losses function),and the error message is InvalidArgumentError (see above for traceback): Input to reshape is a tensor with 11400 values, but the requested shape has 196608.
  2.I try to use session to print the value of multilevel_anchors and multilevel_label_logits in model_fpn.py(multilevel_rpn_losses function),which are input into the rpn_losses.Here is my print process, but there is nothing printed:
with tf.device('/gpu:4'):
            #a = tf.Variable([1,2,3,4,5])
            print('在multilevel_rpn_losses中打印tensor.....................................')
            config = tf.ConfigProto(allow_soft_placement=True)
            config.gpu_options.allow_growth = True
            sess = tf.Session(config=config)
            print(sess.run(multilevel_anchors[0]))
            sess.close()
Thank you for your patience.

**Comment by ppwwyyxx** [comment]
1. As I said: 
For anyone to better understand and diagnose your issue, please post relevant details following the issue template (click "New Issue" -> "Unexpected Problems / Bugs", or visit [this link to post issues about unexpected problems](https://github.com/tensorpack/tensorpack/issues/new?template=unexpected-problems---bugs.md))

2. No that's not how to print things in tensorflow. As I said, to print tensor values during training, the methods are in [the FAQ](https://tensorpack.readthedocs.io/tutorial/faq.html#how-to-print-dump-intermediate-results-in-training). You cannot print tensor values before training because they have no values at that time.

**Comment by chenxiaodanhit** [comment]
### 1. What I did:

(1) This is the command I run: nohup ./train.py --config MODE_MASK=True MODE_FPN=True DATA.BASEDIR=/COCO/DIR BACKBONE.WEIGHTS=ImageNet-R50-AlignPadding.npz &

(2) **If you're using examples, have you made any changes to the examples? Paste `git diff` here:**
I'm using examples/FasterRcnn, and in the code I only changed one line as below: 
(mode_rpn.py in rpn_head() function)
change 
`hidden = Conv2D('conv0', featuremap, channel, 3, activation=tf.nn.relu)`
to
`hidden = Conv2D('conv0', featuremap, channel, 3, dilation_rate=2, activation=tf.nn.relu, padding='SAME')`
### 2. What you observed:

(1) **Include the ENTIRE logs here:**
Here is the logs that I observed that is different from before:
![image](https://user-images.githubusercontent.com/46680345/54485188-4d028c00-48af-11e9-8829-8c2d32624d5e.png)

![image](https://user-images.githubusercontent.com/46680345/54485190-69062d80-48af-11e9-91be-57d8a74aff65.png)

![image](https://user-images.githubusercontent.com/46680345/54485193-6f94a500-48af-11e9-9715-548af7a28973.png)

![image](https://user-images.githubusercontent.com/46680345/54485194-76bbb300-48af-11e9-81d0-caeff48d3c0d.png)

(2) **Other observations, if any:**
I used two GPUs(GTX 1080)
### 3. What you expected, if not obvious.
  What we want is to try to use atrous convolution to see anything change, we konw that atrous convlution has larger recption feild without needing redundant variables.

### 4. Your environment:
  + Python version: 3.6
  + TF version: 1.9
  + Tensorpack version: latest version.
      You can install Tensorpack master by `pip install -U git+https://github.com/ppwwyyxx/tensorpack.git`
      and see if your issue is already solved.
  + If you're not using tensorpack under a normal command line shell (e.g.,
    using an IDE or jupyter notebook), please retry under a normal command line shell. 
  + Hardware information: two GPUS
You may often want to provide extra information related to your issue, but
at the minimum please try to provide the above information __accurately__ to save effort in the investigation.

**Comment by ppwwyyxx** [comment]
This is a tensorflow bug and is now reported at https://github.com/tensorflow/tensorflow/issues/26797 .

**Comment by ppwwyyxx** [comment]
Closing as this is not a tensorpack issue. Feel free to reopen if you have more questions.

**Comment by ppwwyyxx** [comment]
Latest tensorpack (353cd04fe81ea8920bc67f702bf492174b000768) now avoids the bug in `tf.layers`. If you use latest tensorpack, dilated conv should work now.

**Comment by chenxiaodanhit** [comment]
Thank you very much for your instruction!
